src/lib/include/x86_basemath.h

```diff
@@ -0,0 +1,161 @@
+  template<> struct IsVector<__m256d> { static const bool value = true; };
+#endif
+template<typename T>
+using EnableIfVector = typename std::enable_if<IsVector<T>::value, int>::type;
+
+template<typename T, EnableIfVector<T> = 0>
+inline T Abs(T x) {
+  // -0.0 has sign bit 1 and 0 elsewhere. ANDN thus zeroes the sign bit.
+  return andnot_p(set1_p<T>(-0.0), x);
+}
+
+template<typename T, EnableIfVector<T> = 0>
+inline T Nearbyint(T x) {
+  return round_p<_MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC>(x);
+}
+
+template<typename T, EnableIfVector<T> = 0>
+inline T Sqrt(T x) { return sqrt_p(x); }
+
+#ifdef DISTOPIA_X86_AVX2_FMA
+  // Fast version with FMA intrinsics.
+  template<typename T, EnableIfVector<T> = 0>
+  inline T FastMulAdd(T a, T b, T c) { return fmadd_p(a, b, c); }
+  template<typename T, EnableIfVector<T> = 0>
+  inline T FastMulSub(T a, T b, T c) { return fmsub_p(a, b, c); }
+  template<typename T, EnableIfVector<T> = 0>
+  inline T FastNegMulAdd(T a, T b, T c) { return fnmadd_p(a, b, c); }
+  template<typename T, EnableIfVector<T> = 0>
+  inline T FastNegMulSub(T a, T b, T c) { return fnmsub_p(a, b, c); }
+#else
+  // Slower (and less accurate) version for CPUs without FMA.
+  template<typename T, EnableIfVector<T> = 0>
+  inline T FastMulAdd(T a, T b, T c) { return a * b + c; }
+  template<typename T, EnableIfVector<T> = 0>
+  inline T FastMulSub(T a, T b, T c) { return a * b - c; }
+  template<typename T, EnableIfVector<T> = 0>
+  inline T FastNegMulAdd(T a, T b, T c) { return -(a * b) + c; }
+  template<typename T, EnableIfVector<T> = 0>
+  inline T FastNegMulSub(T a, T b, T c) { return -(a * b) - c; }
+#endif
+
+template<typename T, EnableIfVector<T> = 0>
+inline T Remainder(T x, T y) {
+  // FIXME: I suspect wraps_around can be affected by rounding error in ry. Is
+  // this a problem?
+  T wraps_around =
+    round_p<_MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC>(x / y);
+  // FIXME: FastNegMulAdd is inexact on CPUs without FMA.
+  return FastNegMulAdd(wraps_around, y, x);
+}
+
+template<typename T, EnableIfVector<T> = 0>
+inline T FastMin(T x, T y) { return min_p(x, y); }
+
+template<typename T, EnableIfVector<T> = 0>
+inline T DistanceModulo(T x0, T x1, T y) {
+  T d = Abs(x0 - x1);
+  T y_sub_d = y - d;
+  #ifdef DISTOPIA_X86_AVX
+    bool msb_all_zero = testz_p(y_sub_d, y_sub_d);
+  #else
+    // movemask_p(y_sub_d) is a bitfield of sign bits. It is 0 iff all the sign
+    // bits are 0.
+    bool msb_all_zero = !movemask_p(y_sub_d);
+  #endif
+  if (distopia_likely(msb_all_zero)) {
+    return FastMin(d, y_sub_d);
+  }
+  x0 = Remainder(x0, y);
+  x1 = Remainder(x1, y);
+  d = Abs(x0 - x1);
+  return FastMin(d, y - d);
+}
+
+} // namespace
+
+#endif // DISTOPIA_X86_SSE4_1
+
+// 80-bit extended precision handling.
+namespace {
+#if defined(DISTOPIA_X86) && \
+    (defined(DISTOPIA_CLANG) || defined(DISTOPIA_ICC)) && \
+    (__SIZEOF_LONG_DOUBLE__ == 10 || __SIZEOF_LONG_DOUBLE__ == 12 || \
```

Comment:
can this be done with
```c++
constexpr int size =  sizeof(long double)
```
 in case the macro isn't defined for some bizarre reason?

Reply:
That would be nicer but it doesn’t work because `constexpr`s are evaluated after the C preprocessor runs 😕

But this reminds me that this guard is broken anyway and I need to fix it!

Reply:
If the macros aren't defined, then the predicate will evaluate to false and we’ll use the default (slower) implementation.

Reply:
👍